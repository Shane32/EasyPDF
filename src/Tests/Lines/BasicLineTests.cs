using System.Drawing;
using Font = Shane32.EasyPDF.Font;

namespace Tests.Lines;

public class BasicLineTests
{
    private readonly PDFWriter _writer = new PDFWriter();

    public BasicLineTests()
    {
        _writer.ScaleMode = ScaleModes.Inches;
        _writer.NewPage(PageKind.Letter, false, 1f, 1f);
        _writer.PrepForTests();
    }

    [Fact]
    public void FillStyles()
    {
        _writer.FillColor.ShouldBe(Color.Black);

        DrawGeometricFigure(0.5f, false, false, false, false, 0.1f, LineJoinStyle.Miter, 0, LineCapStyle.None);

        _writer.MoveTo(3.75f, 0);
        DrawGeometricFigure(0.5f, true, true, false, false, 0.1f, LineJoinStyle.Miter, 0, LineCapStyle.None);

        _writer.MoveTo(0, 2);
        DrawGeometricFigure(0.5f, true, false, true, false, 0.1f, LineJoinStyle.Miter, 0, LineCapStyle.None);

        _writer.MoveTo(3.75f, 2);
        DrawGeometricFigure(0.5f, true, false, true, true, 0.1f, LineJoinStyle.Miter, 0, LineCapStyle.None);

        _writer.MoveTo(0, 4);
        DrawGeometricFigure(0.5f, true, true, true, true, 0.1f, LineJoinStyle.Miter, 0, LineCapStyle.None);

        _writer.MoveTo(3.75f, 4);
        DrawGeometricFigure(0.5f, true, true, true, false, 0.1f, LineJoinStyle.Miter, 0, LineCapStyle.None);

        _writer.MoveTo(0, 6);
        DrawGeometricFigure(0.5f, true, false, false, false, 0.1f, LineJoinStyle.Miter, 0, LineCapStyle.None);

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }


    [Fact]
    public void LineStyles()
    {
        _writer.ForeColor.ShouldBe(Color.Black);
        _writer.LineStyle.JoinStyle.ShouldBe(LineJoinStyle.Miter);
        _writer.LineStyle.CapStyle.ShouldBe(LineCapStyle.None);
        _writer.LineStyle.DashStyle.ShouldBe(LineDashStyle.Solid);
        _writer.LineStyle.Width.ShouldBe(0.1f, 0.001f);

        _writer.MoveTo(0, 0f);
        DrawGeometricFigure(0.5f, false, true, false, false, 0.1f, LineJoinStyle.Miter, 0, LineCapStyle.None);
        _writer.MoveTo(3.75f, 0f);
        DrawGeometricFigure(0.5f, true, true, false, false, 0.1f, LineJoinStyle.Miter, 0, LineCapStyle.None);

        _writer.MoveTo(0, 2f);
        DrawGeometricFigure(0.5f, false, true, false, false, 0.0625f * 72, LineJoinStyle.Miter, 0, LineCapStyle.None);
        _writer.MoveTo(3.75f, 2f);
        DrawGeometricFigure(0.5f, true, true, false, false, 0.0625f * 72, LineJoinStyle.Miter, 0, LineCapStyle.None);

        _writer.MoveTo(0, 4f);
        DrawGeometricFigure(0.5f, false, true, false, false, 0.0625f * 72, LineJoinStyle.Rounded, 0, LineCapStyle.Round);
        _writer.MoveTo(3.75f, 4f);
        DrawGeometricFigure(0.5f, true, true, false, false, 0.0625f * 72, LineJoinStyle.Rounded, 0, LineCapStyle.Round);

        _writer.MoveTo(0, 6f);
        DrawGeometricFigure(0.5f, false, true, false, false, 0.0625f * 72, LineJoinStyle.Bevel, 0, LineCapStyle.Square);
        _writer.MoveTo(3.75f, 6f);
        DrawGeometricFigure(0.5f, true, true, false, false, 0.0625f * 72, LineJoinStyle.Bevel, 0, LineCapStyle.Square);

        _writer.LineStyle.Width = 0.1f;
        _writer.MoveTo(0, 0f);
        _writer.ForeColor = Color.Blue;
        _writer.LineTo(0, 8f);

        _writer.NewPage(PageKind.Letter, false, 1f, 1f);

        _writer.MoveTo(0, 0f);
        DrawGeometricFigure(0.5f, false, true, false, false, 0.02f * 72, LineJoinStyle.Miter, 0, LineCapStyle.None);
        _writer.MoveTo(3.75f, 0f);
        DrawGeometricFigure(0.5f, false, true, false, false, 0.02f * 72, LineJoinStyle.Miter, 1, LineCapStyle.None);

        _writer.MoveTo(0, 2f);
        DrawGeometricFigure(0.5f, false, true, false, false, 0.02f * 72, LineJoinStyle.Miter, 2, LineCapStyle.None);
        _writer.MoveTo(3.75f, 2f);
        DrawGeometricFigure(0.5f, false, true, false, false, 0.02f * 72, LineJoinStyle.Miter, 3, LineCapStyle.None);

        _writer.MoveTo(0, 4f);
        DrawGeometricFigure(0.5f, false, true, false, false, 0.02f * 72, LineJoinStyle.Miter, 4, LineCapStyle.None);

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    private void DrawGeometricFigure(float scale, bool close, bool border, bool fill, bool eofill, float lineWidth, LineJoinStyle joinStyle, int dashStyle, LineCapStyle capStyle)
    {
        var pos = _writer.Position;
        var len = 0.25f * scale;
        var leg = (float)Math.Sqrt(Math.Pow(len, 2) / 2);

        _writer.ForeColor = Color.Red;
        _writer.LineStyle.Width = lineWidth;
        _writer.LineStyle.JoinStyle = joinStyle;
        if (dashStyle == 0) {
            _writer.LineStyle.DashStyle = LineDashStyle.Solid;
        } else if (dashStyle == 1) {
            _writer.LineStyle.DashStyle = LineDashStyle.Dash;
        } else if (dashStyle == 2) {
            _writer.LineStyle.DashStyle = LineDashStyle.DashDotDot;
        } else if (dashStyle == 3) {
            _writer.LineStyle.DashStyle = LineDashStyle.Dot;
        } else if (dashStyle == 4) {
            _writer.LineStyle.DashStyle = new LineDashStyle(new float[] { 1f, 1f, 2f, 1f, 3f, 5f }, 5f);
        }
        _writer.LineStyle.CapStyle = capStyle;

        _writer
            .LineTo(1f * scale, 0f)
            .CornerTo(.25f * scale, .25f * scale, false)
            .LineTo(0f, 0.5f * scale)
            .BezierTo(0f, 0.25f * scale, 0.25f * scale, 0.25f * scale)
            .LineTo(0.5f * scale, 0)
            .BezierTo(0.25f * scale, 0f, 0.25f * scale + leg, leg)
            .LineTo(0.5f * scale, 0.5f * scale)
            .BezierTo(.25f * scale, .25f * scale, .75f * scale, -0.25f * scale, 1f * scale, 0f)
            .LineTo(0f, 1f * scale)
            .LineTo(-2f * scale, 0f)
            .LineTo(0f, .5f * scale)
            .LineTo(.5f * scale, 0f)
            .LineTo(0f, -1f * scale)
            .LineTo(0.5f * scale, 0f)
            .LineTo(0f, 1.5f * scale)
            .LineTo(-1f * scale, 0f)
            .CornerTo(-0.25f * scale, -0.25f * scale, true);

        if (close) {
            _writer.FinishPolygon(border, fill, eofill);
            _writer.Position.ShouldBe(pos);
        } else {
            _writer.FinishLine();
            _writer.Position.ShouldNotBe(pos);
        }

        _writer.MoveTo(pos);
        _writer.OffsetTo(4.5f * scale, 0).Rectangle(1f * scale, 0.4f * scale, 0, fill, border);
        _writer.OffsetTo(0, 0.7f * scale).Rectangle(1f * scale, 0.5f * scale, 0.2f * scale, fill, border);
        _writer.OffsetTo(0, 0.8f * scale).RectangleDualOffset(1f * scale, 0.7f * scale, 0.15f * scale, 0);
        _writer.OffsetTo(0, 1f * scale).RectangleDualOffset(1f * scale, 0.7f * scale, 0.15f * scale, 0.05f);
    }

    [Fact]
    public void LineDashStylesCompare()
    {
        var style1 = new LineDashStyle(new float[] { 1f, 1f, 2f, 1f, 3f, 5f }, 5f);
        style1.Array.ShouldBe(new float[] { 1f, 1f, 2f, 1f, 3f, 5f });
        style1.Array[0] = 50f;
        style1.Array.ShouldBe(new float[] { 1f, 1f, 2f, 1f, 3f, 5f });
        var style2 = new LineDashStyle(new float[] { 1f, 1f, 2f, 1f, 3f, 5f }, 5f);
        (style1 == style2).ShouldBeTrue();
        style1.Equals(style2).ShouldBeTrue();
        style2.Equals(style1).ShouldBeTrue();
        var style3 = new LineDashStyle(new float[] { 1f, 1f, 2f, 1f, 3f, 5f }, 6f);
        (style1 == style3).ShouldBeFalse();
        style1.Equals(style3).ShouldBeFalse();
        style3.Equals(style1).ShouldBeFalse();
        var style4 = new LineDashStyle(new float[] { 1f, 1f, 2f, 1f, 3f, 5f, 6f }, 5f);
        (style1 == style4).ShouldBeFalse();
        style1.Equals(style4).ShouldBeFalse();
        style4.Equals(style1).ShouldBeFalse();
    }

    [Fact]
    public void InvalidLineDashStyleThrows()
    {
        Should.Throw<ArgumentNullException>(() => new LineDashStyle(null!, 0f));
    }
}
