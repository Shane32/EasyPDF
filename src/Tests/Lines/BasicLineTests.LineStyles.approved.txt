%PDF-1.4
%\xE2\xE3\xCF\xD3
1 0 obj
<</Length 724/Filter/FlateDecode>>stream
x\x9Cm\x96\xC1n\x141\x0C\x86\xEFy\x8Ay\x01Vq\xE2Lf\xAEH\x08\x89\x1B\xD0\x1B\xE2\xD4JH\xA8=\xD0J\xF0\xFA\xC4\xF6\xEFl\x92\xA9\xDA\xDD\xF8\xDF\xC9n\xD2/_=\xFB'||\x08y\xDFj\xD9\xB7\x87\xA7\xF0\xE9!|\xB4\xC5\xF6\xF3\x81\xB6\x9A\xDAo\xDC\x1E_B\xDC\xBE\xB4\xC7\xEF\x10o\xB4\xFD\x0B?~\xB6\xEBO!\xEA\xBCo\x9FQ\xBC\xFEz\x7F\x1A\x8D\xD3^d\xB1\xB8=\x07\x8E\xB7\xB3\xB6\x8A\xCB\xC6\xB7\x98e<\xB7\xC7\xD0\x86T\xE5r\xD9\xDA\xC4\xC2\xF2\xFC7\xB4\x8D\xB4\xF19\x1C$\xE3Qo\xED\x99\xD3M\xAFQ,R\xEC\xF1\xA6S\x88X\xE3)\xCF\x94uR!L\xF7\x89\x8F\x01\xF1\xDC\xED}\xF6\x869\xE1\xB3\xDA\xBA\xF5\x9DX\x0F\xACh\x1BX"\x96~\x0E9O\xF1\xFB\x9DG\xAA \x12I&0\xC9\x04(R\x08\x15\x19\x15\x8B\x14\xEDcr\x02\x98\xCC \x93\x8B\xA2\xC9eb\x93\xEB\xC4&\x1F\x03\x1B\x8E\x03\x1B\xA6\x89\xA2\xD3\xC8\xF9|'v\x1CXt\xCD\x0E\x04\x9BX\xE2\x9DO\x9C\x01\xBD\xDD\x01\x9D\xA3k\xC4l\xF6H\xE1\xFEH\xDDT!>\xE0\x10\x95l\x16Q%\xD3\x88\x8E(\x1E\xC9\xA0"I\xA1&I\x81\x83=\xF6\xC9\xA5\x14y\x92)\xD1\xB8E:\xCB\xA0\x13\xE6v\x9F\x12\xC7Q\xA15\x96cTh\x8D)M\x0E]r\xED'0\xC5\xC5)\x90\x8A\x8E
^I\x10\x7F\x00KK\xA5\xA5\x95\xE2\xD2\xAAa\x11\xBF\x0C\x98\x08f\xC4\xC40\xBDVff8\xCF\xCE\x0C\x92\x81\x19,\x033x\xD5\x99yvH0\xEB\x92\x1D\x13\xD6\xBEd\xE7\xE4{Ys\xE7\x16gp\x83k\xD4Z\x17I\xEB\xDA\xD2q\x98hR\xB8hRKo:\x13DK\xEDE\x15M\x98Y\xBFj\xD0\xA4a\xB5\xC1:Vb\xB4\xACV\xD8\x01\xE7\x1C'\xD12{\xD30\xD1\x9C\xAD\x89\xE6\xFFg&\x16\xE6v\xD1\xC0\xD9\xCDZ"\xB8\xBBYK\xCC\xFB>\x89\xB5fN\xFD\x00\xA6\xB8\x88\x06R\xD1QA4\x09\xDA\xB1\x0C\x96\x95\x15\xA2\x19.\xAD\x12\xF7\x07:\x99\x12\xD3V&\xAF\x97\x99\x19\x0E\xB33\xC3_\xEC-/\x8E\xCC Vg\xE6\xD9!a\xDA\x9A;\xA6\xF9C{\xEE\x9C|/K\xBEs\x8B3\xB8A\xB4\xD4DK*\x1A\xE7d\xA2I\xE1\xA2I-\xF7\xC4\xBCC4f2\xD1\xB8\x9C&\x1A\xEF\x87\x88&\x83\x8A&\x85\x8A&\x05\xCE\xBB\xF2$\x1A\x9F\xDEIL\xB4\x12i\x10\x8D\x8F<\x88\x86\xB9]\xB4\x92\x8E\xD1\xAC5\xF2>\x9A\xB5F\xF2\xDB\xB2\xED\xE3\x92\xF7~\x00S\\D\x03\xA9\xE8\xA8 \x9A\x04\xBDW\x1A,-\x95\x96V\x8AK\xAB\x86E\x842`"\x9A\x11\x13\xD1\xF4Z\x99\x99\xF9a:3X\x02fp\x02\xCC Vg\x86\xDC!A\xACKvLX\xFB\x92\x9D\x13\xF6r\xC9\x9D[\x9C\xC1\xBD]\xBF\xA2\xD1\xFD\xBBW\xBB\x13V\xC3\xFB\x1F\x9D\xA0\xE9\x07
endstream
endobj
3 0 obj
<</Type/Page/MediaBox[0 0 612 792]/Resources<</ProcSet [/PDF /Text /ImageB /ImageC /ImageI]>>/Contents 1 0 R/Parent 2 0 R>>
endobj
4 0 obj
<</Length 532/Filter/FlateDecode>>stream
x\x9C\x85\x94Mn\x1C!\x10F\xF7\x9C\x82\x0B\x04Q?4\xF46R\x14\xC9\xBB\xC4\xB3\xB3\xBC\x1AK\x91,{\x91,\x92\xEB\xA7\xAA(z\x1A\xC6Qd\x0F\xD4\xD7\xD0\x03z\xBC\xE1g\xF8|\x09\xB4\xC5Z\xB6xy\x09_.\xE1[\x80\x98\xE5\xEF\x13\xC4\x8A\xF2\x9F\xE3\xF5=\xE8\x03\x88\xDF\xBFZ\x91\xE3\xAF\x1F\x01\xE3\x83|^CN\x10\xFF\x84\xA7gy\xFA\xE2o\xDA\xB4\x07\xF9\xBC\x06H\xCC\xB7a\x1D|\xD7\xD5r|\x0B\x9C\xD3^\xA5\xE2\x129e\xD2~\x8F\xD7 \x1DV\x1D.Q&\x16\xD6\xF6w\x90\x9DH\xFF\x16\x1Ah\xDFj\x92\x961\xD9\x18\xE4\xA2\xC5\x96\x93M\x01`\x8B\xBB\xB6@6\xA9\x80\x05\x86\xDB\xC4k\xF0\xB8o\xFD\xBD\xFE\xC2\x9C\xFC\xBBd\xDD\xFAA\xAC\xCDW\xEC\x1BX\xA2/\xFD\x16\x88\xA6\xF8x\xA3\xF4\xD4\xD2\xC6Q\x9Bga@(\x88\xB0:\xA4\xDC)\x118&\x02\xE7\xA4\x85\x82\xD2\xDEHi!\xDFL\xE8\xAC\x88\x1D\x16\x15\xA3Ee\xC2Eu\xC2E\xED\x84\x8B\xF3\x09\x17\xC3\x84\xCB\xE3\x00D\xB4\x7F\x10\x0FB\xBE\xE8\x9A\x07#\xDF\xC4\x12o\xC8\xF2\x7F\x98\x19-L\xAD-\xD5\xF0\x0CD;3M\x8B\xE1\x9A\xD6\xA2\x15ps\xDF\xA0P7\x0E*t\xE5\xA0euN;\x93N\x0B\xB3N\x0B\x97\xA0m\x93w\x98y\x12\x0F\xE1\xBCw\xD8\xCBI=\x9F{\xB8\x87\x9C\xCF\xBA\xAD\xB1\xB4\xB3nkD\x9C|\xBB\xCB\xF58\x9A)\x9EY.\xDCT>'\x97\x07:\x17P\x83\x8A\xE6\xF0\xAC4zV\x19>\xAB\x04\x93\x8A\xD8\x01\xAA\x89\x9D\xA0\xAAhcef\xE8\x07\x7F0t\x1B\x9D\xA1\xEB\xE8\x0C]\xC0\x83\xE1\xC8\x03\x9A+x\x97\x076_\xFB.\x0Fnc/k>8\xE6\x7F\x83\xB4{\xCE\x1ACj\x95\x99Y\x93\xB0\x95\xC6\xACD\x192+\xB5\x18Vj\xAD\x97\xDE\x8En%\xCAC\xB3R\x81\xF6\x8BP\x88\xEAM(]\xBF
\x91\xFD.\x94\xA2\xDB@\x94'+\x89\xC7m\xD4\xAD\x1C\xE0\xBB\x95\xE3\xD7\xDA-\xF4\xB9\x87\x95~\x08C\xC3%\xFA\xA1\x0C\x97H\xDB6Y\xB8f\xC6\xE3t\xA6\xF8\x18\xFE\x02\x07\xDB@\xB3
endstream
endobj
5 0 obj
<</Type/Page/MediaBox[0 0 612 792]/Resources<</ProcSet [/PDF /Text /ImageB /ImageC /ImageI]>>/Contents 4 0 R/Parent 2 0 R>>
endobj
2 0 obj
<</Type/Pages/Count 2/Kids[3 0 R 5 0 R]/ITXT(3.3.2.0)>>
endobj
6 0 obj
<</Type/Catalog/Pages 2 0 R>>
endobj
7 0 obj
<<>>
endobj
xref
0 8
0000000000 65535 f 
0000000015 00000 n 
0000001683 00000 n 
0000000806 00000 n 
0000000945 00000 n 
0000001544 00000 n 
0000001754 00000 n 
0000001799 00000 n 
trailer
<</Size 8/Root 6 0 R/Info 7 0 